src/main/webapp/app/pages/companionDiagnosticDevicesPage/companionDiagnosticDevicePage.tsx

```diff
@@ -0,0 +1,633 @@
+        return _.uniq(
+          assoc.alterations.reduce(
+            (acc: any[], alt: any) =>
+              acc.concat(alt.genes.map((gene: any) => gene.hugoSymbol)),
+            []
+          )
+        ).map((gene: any) => ({
+          gene,
+          alterations: _.chain(assoc.alterations)
+            .map((a: any) => a.name)
+            .uniq()
+            .sort()
+            .value(),
+          drugs: assoc.treatments
+            .map((treatment: any) =>
+              treatment.drugs.map((drug: any) => drug.name).join(' + ')
+            )
+            .join(', '),
+          cancerTypes: assoc.associationCancerTypes.reduce(
+            (ctAcc: any[], act: any) => {
+              ctAcc.push(act.cancerType);
+              return ctAcc;
+            },
+            []
+          ),
+          fdaSubmissions: assoc.fdaSubmissions,
+        }));
+      })
+    ).forEach((assoc: IBiomarkerAssociation) => {
+      parsedCompanionDiagnosticDevices.push({
+        name: cdx.name,
+        manufacturer: cdx.manufacturer,
+        platformType: cdx.platformType,
+        specimenTypes: _.chain(cdx.specimenTypes)
+          .map((st: any) => st.name)
+          .sort()
+          .value(),
+        biomarkerAssociation: assoc,
+      });
+    });
+  }
+  return parsedCompanionDiagnosticDevices;
+};
+const parsedCDx: ICompanionDiagnosticDevice[] = parseCDx();
+
+const downloadCDx = () => {
+  const seperator = '\t';
+  const headers = [
+    'Gene',
+    'Alteration(s)',
+    'Cancer Type',
+    'Drug(s)',
+    'Companion Diagnostic Device',
+    'Specimen Types(s)',
+    'Platform Type',
+  ].join(seperator);
+  const rows = parsedCDx.map(cdx => {
+    const row = [];
+    row.push(cdx.biomarkerAssociation.gene);
+    row.push(cdx.biomarkerAssociation.alterations.join(', '));
+    row.push(getAllTumorTypesName(cdx.biomarkerAssociation.cancerTypes));
+    row.push(cdx.biomarkerAssociation.drugs);
+    row.push(`${cdx.name} (${cdx.manufacturer})`);
+    row.push(cdx.specimenTypes?.join(', ') || '');
+    row.push(cdx.platformType || '');
+    return row.join(seperator);
+  });
+
+  downloadFile(
+    'companion_diagnostic_devices.tsv',
+    [headers, ...rows].join('\r\n')
+  );
+};
+
+const CompanionDiagnosticDevicePage: React.FunctionComponent<{}> = () => {
+  const [hasFilter, setHasFilter] = useState(false);
+  const [filteredCDx, setFilteredCdx] = useState<ICompanionDiagnosticDevice[]>(
+    parsedCDx
+  );
+  const [alterationSearch, setAlterationSearch] = useState('');
+  const [selectedGenes, setSelectedGenes] = useState<SelectOption[]>([]);
+  const [
+    relevantCancerTypeSearchKeyword,
+    setRelevantCancerTypeSearchKeyword,
+  ] = useState('');
+  const [selectedDrug, setSelectedDrug] = useState<SelectOption | null>(null);
+  const [selectedCDx, setSelectedCDx] = useState<SelectOption | null>(null);
+
+  useEffect(() => {
+    const filterPresent =
+      !!relevantCancerTypeSearchKeyword ||
```

Comment:
The purpose of adding double exclamation mark seems redundant

Reply:
Its to convert the value to a boolean. Otherwise if the search keyword was "example", then `filterPresent = 'example'`